src/audio/multiband_drc/multiband_drc_generic.c

```diff
@@ -0,0 +1,411 @@
+					   struct audio_stream *sink,
+					   uint32_t frames)
+{
+	int32_t *x;
+	int32_t *y;
+	int i;
+	int n = source->channels * frames;
+
+	for (i = 0; i < n; i++) {
+		x = audio_stream_read_frag_s32(source, i);
+		y = audio_stream_write_frag_s32(sink, i);
+		*y = *x;
+	}
+}
+#endif /* CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE */
+
+static void multiband_drc_process_emp_crossover(struct multiband_drc_state *state,
+						crossover_split split_func,
+						int32_t *buf_src,
+						int32_t *buf_sink,
+						int enable_emp,
+						int nch,
+						int nband)
+{
+	struct iir_state_df2t *emp_s;
+	struct crossover_state *crossover_s;
+	int32_t *buf_sink_band;
+	int ch, band;
+	int32_t emp_out;
+	int32_t crossover_out[nband];
+
+	for (ch = 0; ch < nch; ch++) {
+		emp_s = &state->emphasis[ch];
+		crossover_s = &state->crossover[ch];
+
+		if (enable_emp)
+			emp_out = iir_df2t_2stage_biquads(emp_s, *buf_src);
+		else
+			emp_out = *buf_src;
+
+		split_func(emp_out, crossover_out, crossover_s);
+		buf_sink_band = buf_sink;
+		for (band = 0; band < nband; band++) {
+			*buf_sink_band = crossover_out[band];
+			buf_sink_band += PLATFORM_MAX_CHANNELS;
+		}
+
+		buf_src++;
+		buf_sink++;
+	}
+}
+
+static void multiband_drc_process_drc(struct drc_state *state,
+				      const struct sof_drc_params *p,
+				      int32_t *buf_src,
+				      int32_t *buf_sink,
+				      int nbyte,
+				      int nch)
+{
+	int16_t *pd_write16;
+	int16_t *pd_read16;
+	int32_t *pd_write32;
+	int32_t *pd_read32;
+	int ch;
+	int pd_write_index;
+	int pd_read_index;
+	int is_2byte = (nbyte == 2); /* otherwise is 4-bytes */
+
+	if (p->enabled && !state->processed) {
+		drc_update_envelope(state, p);
+		drc_compress_output(state, p, nbyte, nch);
+		state->processed = 1;
+	}
+
+	pd_write_index = state->pre_delay_write_index;
+	pd_read_index = state->pre_delay_read_index;
+
+	if (is_2byte) { /* 2 bytes per sample */
+		for (ch = 0; ch < nch; ++ch) {
+			pd_write16 = (int16_t *)state->pre_delay_buffers[ch] + pd_write_index;
+			pd_read16 = (int16_t *)state->pre_delay_buffers[ch] + pd_read_index;
+			*pd_write16 = sat_int16(Q_SHIFT_RND(*buf_src, 31, 15));
+			*buf_sink = *pd_read16 << 16;
+
+			buf_src++;
+			buf_sink++;
+		}
+	} else { /* 4 bytes per sample */
+		for (ch = 0; ch < nch; ++ch) {
+			pd_write32 = (int32_t *)state->pre_delay_buffers[ch] + pd_write_index;
+			pd_read32 = (int32_t *)state->pre_delay_buffers[ch] + pd_read_index;
+			*pd_write32 = *buf_src;
+			*buf_sink = *pd_read32;
+
+			buf_src++;
+			buf_sink++;
+		}
+	}
+
+	pd_write_index = (pd_write_index + 1) & DRC_MAX_PRE_DELAY_FRAMES_MASK;
+	pd_read_index = (pd_read_index + 1) & DRC_MAX_PRE_DELAY_FRAMES_MASK;
+	state->pre_delay_write_index = pd_write_index;
+	state->pre_delay_read_index = pd_read_index;
+
+	/* Only perform delay frames by early return here if not enabled */
+	if (!p->enabled)
+		return;
+
+	/* Process the input division (32 frames). */
+	if (!(pd_write_index & DRC_DIVISION_FRAMES_MASK)) {
```

Comment:
just to make sure i understand this, you overflow the mask to see when you have enough samples? Clever, but would make sure that the pre_delay frames is clearly stated to be 2^n

Reply:
Yes, pre_delay frames is stated to be a multiple of DRC_DIVISION_FRAMES
https://github.com/thesofproject/sof/blob/master/src/audio/drc/drc.c#L101